File: content/polympiads.2/2.distance-a-la-racine/.codes/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

int main () {
    int nbArrets;
    cin >> nbArrets;

    vector<int> parents(nbArrets);
    for (int i = 0; i < nbArrets; i ++) {
        cin >> parents[i];
    }

    vector<int> distancesAuxParents (nbArrets);
    for (int i = 0; i < nbArrets; i ++) {
        cin >> distancesAuxParents[i];
    }
    
    // On calcule la distance de la manière suivante, pour tout noeud possédant un parent,
    // La distance au centre est la distance au centre du parent plus la distance au parent
    vector<int> distancesAuCentre (nbArrets);
    for (int iArret = 1; iArret < nbArrets; iArret ++) {
        distancesAuCentre[iArret] = distancesAuCentre[parents[iArret]] + distancesAuxParents[iArret];
    }

    for (int iArret = 0; iArret < nbArrets; iArret ++) {
        cout << distancesAuCentre[iArret];
    }
}

File: content/polympiads.2/2.distance-a-la-racine/.codes/solution.py
nb_arrets = int(input())
parents = list(map(int, input().split()))
distances_aux_parents = list(map(int, input().split()))

# On calcule la distance de la manière suivante, pour tout noeud possédant un parent,
# La distance au centre est la distance au centre du parent plus la distance au parent
distances_au_centre = [0 for i in range(nb_arrets)]
for i_arret in range(1, nb_arrets):
	distances_au_centre[i_arret] = distances_au_centre[parents[i_arret]] + distances_aux_parents[i_arret]

for i_arret in range(nb_arrets):
	print(distances_au_centre[i_arret])